### ISSUE TITLE
Customizable SetDuration for Valve Accessory

### ISSUE THREAD

**Issue by tseiger1** [justjam2013/homebridge-virtual-accessories#64]
Thanks for this amazing plugin—it’s been a real treat for my HomeKit setup!

I have a feature request for the Valve accessory. It would be awesome to have the option to customize the SetDuration characteristic with a minimum, maximum, and step size.

For example, I use a valve to water my balcony planters, and being able to set times e.g. between 1 and 5 minutes (in 1-minute steps) directly in the Home app would be super helpful. 
It would also save me from accidentally turning my balcony into a miniature Venice every time I forget to stop the water :)

Thanks for considering this!

**Comment by justjam2013** [comment]
Thank you for your suggestion, but I don't fully understand what you're asking.

The configuration already has a duration field as you can see from the screenshot below. Per the spec, you can set a maximum duration of 1 hour and the unit is seconds, so 3600 secs max - I need to update the field description on that.
If you set the duration to zero, then you have turn it off manually.

Maybe I need to change the name if it is confusing?
![Screenshot 2024-12-20 at 11 07 08 PM](https://github.com/user-attachments/assets/24b47a8a-2182-42cd-a0af-4b325743cc1b)<p>

**Comment by justjam2013** [comment]
Sorry, I just re-read your request ... let me try again!

In the Home app, when you open the valve accessory settings you see a selection widget to select the valve run time:

![IMG_6596](https://github.com/user-attachments/assets/1e95d91c-6abb-411f-9654-7ea63e60370d)<p>

The available duration times are 15 seconds, 5 minutes, 10 minutes, etc. If I understand correctly, you would like to see the options as 15 seconds, 1 minute, 2 minutes, 3 minutes, 4 minutes, 5 minutes, 10 minutes, etc. Is this correct?

Unfortunately, the plugin has no control over what the Home app displays and the Home app is notorious for providing options that are less than ideal. A suggestion is to install the [Eve app](https://www.evehome.com/en-us/eve-app), it's a free app. It doesn't exactly offer what you are looking for, but it does offer a few additional duration times: 10 seconds, 30 seconds, 1 minute, 2 minutes, 5 minutes, 10 minutes, etc.

![IMG_6598](https://github.com/user-attachments/assets/f3539e0a-38b1-4acc-8fe9-4c2fe7827660)

**Comment by tseiger1** [comment]
Thank you for your response and for the suggestion to use the Eve app - I appreciate the tip!

Yes, I would like to change those options shown in the Home app to include durations like 15 seconds, 1 minute, 2 minutes, 3 minutes, 4 minutes, 5 minutes, 10 minutes, etc. as the defaults of the Home app are truly often not ideal.

I believe the plugin could override these defaults by implementing something like the following:

``` TypeScript 
import { CharacteristicProps, Formats, Perms } from 'homebridge';

const durationProps: CharacteristicProps = {
  format: Formats.UINT32, 
  perms: [Perms.PAIRED_READ, Perms.PAIRED_WRITE, Perms.NOTIFY], 
  minValue: 0,
  maxValue: 300,
  minStep: 60, 
};

this.service.getCharacteristic(this.platform.Characteristic.SetDuration)
  .setProps(durationProps)  
  .onSet(this.handleSetDurationSet.bind(this))
  .onGet(this.handleSetDurationGet.bind(this)); 
```

However, TypeScript isn’t exactly my mother tongue, so this might need some adjustment, but I hope it conveys the idea. 

 This approach could potentially override the default behavior, allowing custom durations.  If there’s a way to integrate such changes into the plugin, it could address these less-than-ideal defaults effectively.

Looking forward to your thoughts on that.

**Comment by justjam2013** [comment]
That's an interesting suggestion.

The code snippet you provided modifies the properties of the SetDuration characteristic, as defined in the Homekit Accessory Protocol:
```
      unit: Units.SECONDS,
      minValue: 0,
      maxValue: 3600,
      minStep: 1,
```

I will definitely look into it, however it may take a little while for me to get to it, as I am working on some other fixes at the moment.

**Comment by tseiger1** [comment]
Awesome! Sure, please take your time

**Comment by justjam2013** [comment]
[Breaking changes in HAP-NodeJS v1](https://github.com/homebridge/homebridge/wiki/Updating-To-Homebridge-v2.0#hap-nodejs-v1)
